dp/drivers/common.c

```diff
@@ -0,0 +1,313 @@
+	return 1;
+}
+
+static int dpdk_devinit(struct pci_dev *pci_dev, struct rte_pci_driver **found_driver)
+{
+	int ret;
+	struct rte_pci_device *dpdk_pci_dev = NULL;
+	struct rte_pci_addr addr;
+	struct rte_pci_driver *driver;
+
+	addr.domain = pci_dev->addr.domain;
+	addr.bus = pci_dev->addr.bus;
+	addr.devid = pci_dev->addr.slot;
+	addr.function = pci_dev->addr.func;
+
+	*found_driver = NULL;
+
+	TAILQ_FOREACH(dpdk_pci_dev, &pci_device_list, next) {
+		if (rte_eal_compare_pci_addr(&dpdk_pci_dev->addr, &addr))
+			continue;
+
+		dpdk_pci_dev->mem_resource[0].addr = pci_map_mem_bar(pci_dev, &pci_dev->bars[0], 0);
+
+		TAILQ_FOREACH(driver, &pci_driver_list, next) {
+			ret = rte_eal_pci_probe_one_driver(driver, dpdk_pci_dev);
+
+			if (ret < 0) {
+				/* negative value is an error */
+				return -1;
+			} else if (ret > 0) {
+				/* positive value means driver not found */
+				continue;
+			}
+
+			/* driver found */
+			*found_driver = driver;
+			return 0;
+		}
+
+		/* driver not found */
+		return -1;
+	}
+
+	/* device not found */
+	return -1;
+}
+
+static enum rte_eth_rx_mq_mode translate_conf_rxmode_mq_mode(enum ix_rte_eth_rx_mq_mode in)
+{
+	switch (in) {
+	case IX_ETH_MQ_RX_RSS:
+		return ETH_MQ_RX_RSS;
+	default:
+		assert(false);
+	}
+}
+
+static enum rte_eth_tx_mq_mode translate_conf_txmode_mq_mode(enum ix_rte_eth_tx_mq_mode in)
+{
+	switch (in) {
+	case IX_ETH_MQ_TX_NONE:
+		return ETH_MQ_TX_NONE;
+	default:
+		assert(false);
+	}
+}
+
+static uint16_t translate_conf_rss_hf(uint16_t in)
+{
+	uint16_t out = 0;
+
+#define COPY_AND_RESET(dst, src, bit_out, bit_in) \
+	if ((src) & (bit_in)) { \
+		(dst) |= (bit_out); \
+		(src) &= ~(bit_in); \
+	}
+
+	COPY_AND_RESET(out, in, ETH_RSS_NONFRAG_IPV4_TCP, ETH_RSS_IPV4_TCP);
+	COPY_AND_RESET(out, in, ETH_RSS_NONFRAG_IPV4_UDP, ETH_RSS_IPV4_UDP);
+
+#undef COPY_AND_RESET
+
+	assert(in == 0);
+
+	return out;
+}
+
+static void translate_conf(struct rte_eth_conf *out, struct ix_rte_eth_conf *in)
+{
+	char *p;
+
+	memset(out, 0, sizeof(struct rte_eth_conf));
+
+#define COPY_AND_RESET(field) \
+	do { \
+		out->field = in->field; \
+		in->field = 0; \
+	} while (0)
+
+#define COPY_AND_RESET2(field, f) \
+	do { \
+		out->field = f(in->field); \
+		in->field = 0; \
+	} while (0)
+
+	COPY_AND_RESET2(rx_adv_conf.rss_conf.rss_hf, translate_conf_rss_hf);
+	COPY_AND_RESET(rxmode.header_split);
+	COPY_AND_RESET(rxmode.hw_ip_checksum);
+	COPY_AND_RESET(rxmode.hw_strip_crc);
+	COPY_AND_RESET(rxmode.hw_vlan_filter);
+	COPY_AND_RESET(rxmode.jumbo_frame);
+	COPY_AND_RESET2(rxmode.mq_mode, translate_conf_rxmode_mq_mode);
+	COPY_AND_RESET(rxmode.split_hdr_size);
+	COPY_AND_RESET2(txmode.mq_mode, translate_conf_txmode_mq_mode);
+
+#undef COPY_AND_RESET
+#undef COPY_AND_RESET2
+
+	for (p = (char *) in; p < (char *) in + sizeof(*in); p++)
+		assert(!*p);
+}
+
+extern int ixgbe_init(struct ix_rte_eth_dev *dev, struct rte_pci_driver *driver);
+
+struct rte_eth_rxconf rx_conf;
```

Comment:
I agree that you should use them in i40e. You should put declarations in a header file and definition in a source file, e.g.: `struct rte_eth_rxconf rx_conf` is a definition so it should stay in common.c since it is common for all drivers.